lib/android.js

```diff
@@ -1378,16 +1378,53 @@ function makeArtQuickFrameInfoGetter (impl) {
 }
 
 function _getArtQuickFrameInfoGetterThunk (impl) {
-  const thunk = makeThunk(64, writer => {
-    writer.putPushRegReg('x0', 'lr');
-    writer.putCallAddressWithArguments(impl, ['x1']);
-    writer.putPopRegReg('x2', 'lr');
-    writer.putStrRegRegOffset('x0', 'x2', 0);
-    writer.putStrRegRegOffset('w1', 'x2', 8);
-    writer.putRet();
-  });
+  if (Process.arch == 'arm64') {
+    const thunk = makeThunk(64, writer => {
+      writer.putPushRegReg('x0', 'lr');
+      writer.putCallAddressWithArguments(impl, ['x1']);
+      writer.putPopRegReg('x2', 'lr');
+      writer.putStrRegRegOffset('x0', 'x2', 0);
+      writer.putStrRegRegOffset('w1', 'x2', 8);
+      writer.putRet();
+    });
+    return new NativeFunction(thunk, 'void', ['pointer', 'pointer'], nativeFunctionOptions);
+  } else if (Process.arch == 'arm') {
+    const thunk = makeThunk(64, writer => {
+      // By call convention, we pass a pointer for the result struct
+      writer.putCallAddressWithArguments(impl, ['r0', 'r1']);
+      writer.putPopRegs(['r0', 'lr']);
+      writer.putMovRegReg('pc', 'lr');
+    });
+    return new NativeFunction(thunk, 'void', ['pointer', 'pointer'], nativeFunctionOptions);
+  } else if (Process.arch == 'ia32') {
+    const thunk = makeThunk(64, writer => {
+      writer.putMovRegRegOffsetPtr('ecx', 'esp', 4); // result
+      writer.putMovRegRegOffsetPtr('edx', 'esp', 8); // self
+      writer.putCallAddressWithArguments(impl, ['ecx', 'edx']);
+      
+      // restore callee's stack frame
+      writer.putMovRegReg('esp', 'ebp');
+      writer.putPopReg('ebp');
+
+      writer.putRet();
+    });
+    return new NativeFunction(thunk, 'void', ['pointer', 'pointer'], nativeFunctionOptions);
+  } else if (Process.arch == 'x64') {
+    const thunk = makeThunk(64, writer => {
+
+      writer.putPushReg('rdi') // preserve result buffer pointer
+      writer.putCallAddressWithArguments(impl, ['rsi']); // self
+      writer.putPopReg('rdi');
 
-  return new NativeFunction(thunk, 'void', ['pointer', 'pointer'], nativeFunctionOptions);
+      // Struct is stored by value in the rax and edx register
+      // Write struct to result buffer
+      writer.putMovRegPtrReg('rdi', 'rax');
+      writer.putMovRegOffsetPtrReg('rdi', 8, 'edx');
+
+      writer.putRet();
+    });
+    return new NativeFunction(thunk, 'void', ['pointer', 'pointer'], nativeFunctionOptions);
+  }
```

Comment:
- All of these thunks are 64 bytes large – can we reduce some of them? (Previously I've looked at how much code is generated, and rounded up to the nearest power-of-two.)
- Would be good to use `switch` to avoid repeatedly reading `Process.arch`
- Seems the `NativeFunction` construction could be moved out of the architecture-specific code
- For consistency the architectures should be in the following order: ia32, x64, arm, arm64

Reply:
I reduced the thunk size now. I suppose the nearest (rounded up) power-of-two is still preferable?